concourse/pipelines/guest-package-build.jsonnet

```diff
@@ -0,0 +1,1177 @@
+local underscore(input) = std.strReplace(input, '-', '_');
+
+local buildpackagejob = {
+  local tl = self,
```

Comment:
This is a pretty big review :)

For review's sake, I'd feel a lot more comfortable if this were split in two:
1. A port of the previous YAML file
2. Improvements / enhancements

In the PR's overview you mentioned that you weren't able to get the same YAML output as in previous reviews. What kind of challenges were there?

Reply:
we had some lengthy discussion offline; my position is it's not practical to make it smaller without adding a lot more work.

in previous translations i was able to guarantee the output was identical because we'd already cleaned up the yaml versions significantly and because the jobs such as image build jobs were naturally more uniform. by contrast, the package build pipeline is more or less the same as the original version we checked in, and package builds have a wide amount of variations - architecture, destination, naming, etc. to make templating work and add value i had to accept that it would also involve changes, such as moving the upload steps into the build jobs, adopting the new promotion RPCs to replace deprecated ones, and unifying some GCS paths. to make it templatized but not make these changes would have been a much more complicated set of templates, which would themselves then go away shortly thereafter. it would more or less double the work involved.